src/hostname/hostnamectl.c

```diff
@@ -73,7 +76,68 @@ static const char* chassis_string_to_glyph(const char *chassis) {
         return NULL;
 }
 
-static int print_status_info(StatusInfo *i) {
+static const char* const all_fields[] = {
+                "architecture",
+                "chassis",
+                "deployment",
+                "hardware-model",
+                "hardware-vendor",
+                "home-url",
+                "hostname",
+                "icon-name",
+                "kernel-name",
+                "kernel-release",
+                "location",
+                "os-cpe-name",
+                "os-pretty-name",
+                "pretty-hostname",
+                "static-hostname",
+                "virtualization",
+                NULL};
+
+static int show_fields(const StatusInfo *i, char **fields) {
+        assert(i);
+
+        STRV_FOREACH(field, fields)
+                if (streq(*field, "hostname"))
+                        printf("%s\n", strempty(i->hostname));
+                else if (streq(*field, "static-hostname"))
+                        printf("%s\n", strempty(i->static_hostname));
+                else if (streq(*field, "pretty-hostname"))
+                        printf("%s\n", strempty(i->pretty_hostname));
+                else if (streq(*field, "icon-name"))
+                        printf("%s\n", strempty(i->icon_name));
+                else if (streq(*field, "chassis"))
+                        printf("%s\n", strempty(i->chassis));
+                else if (streq(*field, "deployment"))
+                        printf("%s\n", strempty(i->deployment));
+                else if (streq(*field, "location"))
+                        printf("%s\n", strempty(i->location));
+                else if (streq(*field, "kernel-name"))
+                        printf("%s\n", strempty(i->kernel_name));
+                else if (streq(*field, "kernel-release"))
+                        printf("%s\n", strempty(i->kernel_release));
+                else if (streq(*field, "os-pretty-name"))
+                        printf("%s\n", strempty(i->os_pretty_name));
+                else if (streq(*field, "os-cpe-name"))
+                        printf("%s\n", strempty(i->os_cpe_name));
+                else if (streq(*field, "virtualization"))
+                        printf("%s\n", strempty(i->virtualization));
+                else if (streq(*field, "architecture"))
+                        printf("%s\n", strempty(i->architecture));
+                else if (streq(*field, "home-url"))
+                        printf("%s\n", strempty(i->home_url));
+                else if (streq(*field, "hardware-vendor"))
+                        printf("%s\n", strempty(i->hardware_vendor));
+                else if (streq(*field, "hardware-model"))
+                        printf("%s\n", strempty(i->hardware_model));
+                else
+                        assert_not_reached();
+
+        return 0;
+}
+
```

Comment:
Maybe something like the following would avoid having to repeat the list of the known properties multiple times.

```C
typedef struct StatuInfoField {
     const char *name;
     size_t pos;
} StatusInfoField;

StatusInfoField status_info_fields[] = {
     { "hostname", offsetof(StatusInfo, hostname) },
     ...
     { }
};

static int show_fields(const StatusInfo *i, char **fields) {
     assert(i);
     
     STRV_FOREACH(field, fields)
          for (StatusInfoField *field = status_info_fields; field->name; field++)
               if (streq(field->name, name))
                    printf("%s\n", *(const char **)((void *)i + field->pos));

     assert_not_reached();

     return 0;
}
```
Generating `StatusInfo` and the `StatusInfoFileds` array with a single macro is left as an exercise :)

Reply:
Thanks. Updated like this.

Reply:
It does not matter much, but I think that the following may look a bit more idiomatic:
```c
static int show_fields(const StatusInfo *info, char **names) {
        assert(info);

        STRV_FOREACH(name, names)
                for (const StatusInfoField *field = status_info_fields; field->name; field++)
                        if (streq(*name, field->name))
                                return printf("%s\n", *(const char **)((uint8_t*) info + field->offset));

        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Unknown property: %s", *name);
}
```

Reply:
This would work if there was only one field. With multiple items in names, we can't invert the check like this.